### PR TITLE
LF: retire VersionTimeline

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This is part of #7788

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 1]
```diff
@@ -69,10 +67,10 @@ class DecodeV1Spec
     forEveryVersionSuchThat(_ => true)
 
   private[this] def forEveryVersionBefore[U](maxVersion: LV): (LV => U) => Unit =
-    forEveryVersionSuchThat(v => v precedes maxVersion)
+    forEveryVersionSuchThat(_ < maxVersion)
```
Comment (hurryabit) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:70]: Maybe we should inline this function and the one below given how simple it is to write the version constraints now. I feel like these functions only add complexity without any gain in the new situation.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/VersionRange.scala [hunk 2]
```diff
@@ -22,24 +20,21 @@ import com.daml.lf.transaction.VersionTimeline
...
     VersionRange(
-      min = maxVersion(this.min, that.min),
-      max = minVersion(this.max, that.max)
+      min = this.min max that.min,
+      max = this.max min that.max,
     )
 
   def isEmpty: Boolean =
-    max precedes min
+    ordering.lt(max, min)
```
Comment (hurryabit) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/VersionRange.scala:32]: It seems somewhat inconsistent to not write `max < min` here like everywhere else. 😄

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 3]
```diff
@@ -197,12 +196,15 @@ object TransactionCoder {
...
+            EncodeError(node.version, isTooOldFor = "non-empty choice-observers")
           for {
             _ <- Either.cond(
-              test =
-                !((node.version precedes TransactionVersions.minChoiceObservers) && ne.choiceObservers.nonEmpty),
+              test = exe.version >= TransactionVersion.minChoiceObservers ||
+                exe.choiceObservers.isEmpty,
               right = (),
-              left = EncodeError(node.version, isTooOldFor = "non-empty choice-observers")
+              left = x
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:207]: Aren't we losing laziness here by binding the error to a variable first? Is that intended? Is this a performance critical code path we should worry about?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 4]
```diff
@@ -197,12 +196,15 @@ object TransactionCoder {
           }
 
         case ne @ NodeExercises(_, _, _, _, _, _, _, _, _, _, _, _, _, _, _) =>
+          def exe = ne
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:199]:
```suggestion
        case exe @ NodeExercises(_, _, _, _, _, _, _, _, _, _, _, _, _, _, _) =>
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala [hunk 5]
```diff
@@ -28,8 +28,10 @@ object ValueCoder {
   final case class DecodeError(errorMessage: String)
 
   object DecodeError extends (String => DecodeError) {
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala:30]: This looks strange to me. But I guess it's just a Scala idiom I don't know.

**Review by hurryabit — APPROVED** [review]
Much nicer. I'm also very happy I don't need to break my head about the meaning of `!(x precedes y)` anymore. Thanks a lot.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/VersionRange.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/VersionRange.scala:32]: dropped

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala]: done.

**Pull request merged**